### DISCUSSION TITLE
macOS M1 Support

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#6844] [General]
I'm would like to switch to using MacBook Pros for development at some point in the future. (On my end, there is no reason to stay with Windows as UWP development is going to be phased out after switching to Avalonia/Uno exclusively). MacBook Pros are nicer hardware and better for mobile development in my opinion -- if only the software was up to the task. I've got to say the 14" M1 Max is amazing at everything I throw at it that runs -- best all-around laptop I've seen. (I'm sure some will disagree)

That said, what is the status/plan for M1 support with Avalonia?

 1. I expect you are waiting for .NET 6 RTM to provide full support. I've also seen some initial experimentation PRs for .NET 5/6. Is there high confidence .NET 6 has everything you need to make this work? (It would be great not to have to wait another year for .NET 7.
 2. Has anyone been testing Avalonia development with the M1? Any findings so far in terms of issues or does it look like once VS 2022 / Rider vNext Stable comes along everything will work?
 3. With Rider specifically, how is that working? Any issues other than lack of debugging support?
 4. Will it be possible to build Avalonia itself on M1 hardware? (I seemed to read in the docs previously that Avalonia can only be built on Windows even today).

I'll get around to confirming some of these myself at some point. Thought I would ask first and see if anyone has any info though.

**Comment by robloo** [comment]
Off topic for Avalonia but someone might find it useful:

In some early experimentation with the new M1 Max MacBook Pros I of course ran across a myriad of issues and incompatibilities. These issues were primarily encountered while trying to do a hybrid Windows/Mac development environment like you could do on the old Intel macs. It just doesn't seem possible with the M1's yet though and likely won't be for several years (if ever). Windows needs much better arm support or an x86 emulator with Rosetta 2 performance is needed to run Windows.

Anyway, what I found so far: (for anyone attempting something similar)
 - Windows 10 ARM virtualization with Parallels is high performance but very poor compatiblity. A lot of software has issue running here and even Visual Studio isn't officially support but does run. Most higher performance apps will have stability issues.
 - For any ex-UWP developers, Windows 10 ARM does NOT allow building apps. The issue is there is no ARM compatible tooling to compile and debug apps even though Visual Studio runs and UWP has ARM binaries for each app.
 - The Android emulator does NOT appear to work yet on M1 macs. This is an issue for anyone doing mobile apps with either Uno or Avalonia -- debugging with a physical device seems to work as a work-around.
 - Visual Studio 2022 for mac isn't ready yet and M1 support is pushed back again but they say it will still be ready for M1 in the .NET 6 RTM.
 - All things Apple work great of course -- but .NET support for M1 is needed quickly in both Visual Studio and the framework.
 - x86 emulation of Windows 10 with UTM is actually possible but not usable due to performance.
 - Wine helps running some things like games, but isn't useful for developers. Parallels is the only option but then you are faced with the Windows ARM incompatibilities mentioned above.

If anyone has other questions about this scenario do let me know.

**Comment by kekekeks** [comment]
> Is there high confidence .NET 6 has everything you need to make this work? (It would be great not to have to wait another year for .NET

We also need updated SkiaSharp. `libSkiaSharp.dylib` from the latest _stable_ release still only targets x86_64. We could switch to 2.88 _preview_ releases, but that's also a breaking change.

>Has anyone been testing Avalonia development with the M1? Any findings so far in terms of issues or does it look like once VS 2022 / Rider vNext Stable comes along everything will work?

We've tried to building our native backend and it seems to work. Not quite sure how we'll be doing packaging (fat binary or multiple builds)

> With Rider specifically, how is that working? Any issues other than lack of debugging support?

No info on that one yet, but shouldn't be different from any other .NET 6 app.

> Will it be possible to build Avalonia itself on M1 hardware?  (I seemed to read in the docs previously that Avalonia can only be built on Windows even today).

WPF and WinForms interop libraries require a windows machine to build. libAvaloniaNative.dylib requires a macOS machine with XCode to build. Android and iOS backends require either Windows or macOS machine to build. So packaging process is a bit complicated, yes. However if you only need to build libraries for the platform you are currently debugging you can just clone and build.

The only problem with M1 might be running our Nuke script (builds nuget packages on CI) since it requires .NET Core 3.1 runtime, but that one seems to be running via Rosetta just fine.

**Reply by robloo** [reply]
Thanks for all the info! Glad to hear there is lots of investigation already completed for M1.

> WPF and WinForms interop libraries require a windows machine to build. libAvaloniaNative.dylib requires a macOS machine with XCode to build. Android and iOS backends require either Windows or macOS machine to build. So packaging process is a bit complicated, yes. However if you only need to build libraries for the platform you are currently debugging you can just clone and build.

Great to know. My use case here would be only to build Avalonia for macOS on M1. That would be for PRs and changes to Avalonia itself for new features/testing. I should NOT need to cross-compile any builds of Avalonia (or apps for that matter). I just wanted to make sure I could clone the repo, run the sample app and make changes all on the M1. Sounds like that is all good to go (once the known updates are released).

**Reply by jmacato** [reply]
Idk if this is of any use but me and @danwalmsley are using Avalonia repo in m1 since last year and we barely had any problems with it (thanks to rosetta 😃)

**Comment by BAndysc** [comment]
I am using M1 mac for quite some time now and I can confirm it works pretty good (not perfect tho). Since June I am using .NET 6 preview + Rider for Apple Silicon.

As kekekeks mentioned Avalonia is using stable SkiaSharp which doesn't yet support M1, but preview versions do and luckily, it was enough to include preview packages in the csproj
```
        <PackageReference Include="HarfBuzzSharp" Version="2.8.2-preview.127" />
        <PackageReference Include="SkiaSharp" Version="2.88.0-preview.127" />
``` 
And those packages override default Avalonia's dependencies dlls, so that I can run natively Avalonia app on M1 with .NET 6 and I have absolutely no problem with it.

Occasionally rider likes to crash while debugging .net app on M1, but it doesn't happen _too_ often. 

Edit: also, as of now, you are forced to use EAP version of Rider, which can have more bugs, for instance latest version crashes on my solution (but after sending an issue, it turned out this will be fixed in the next preview version).

**Reply by robloo** [reply]
I'm glad you've had good luck running Rider on the M1 after they released support for it (in preview I realize). The preview packages don't worry me so much -- I was just looking to know if this was possible at this point in time without too much trouble. I know nothing worked at the beginning of the year. To my surprise it seems relatively straightforward now. Thanks for confirming how you have setup your dev environment to do this!

I'll probably download the latest trial of Rider and give this a try myself in a few days. Thanks for the trick with the preview packages, I'm sure that saved me a lot of time to track down.

**Reply by robloo** [reply]
I installed Rider EAP (clicking the .NET core developer tooling option) and then cloned an Avalonia app to test. It worked the very first try! I had forgotten how nice it was to work with a functional/updated IDE and tooling...

So it seems like this works just fine but I will test more later. Really glad to see Avalonia development on M1 is working this well.

**Comment by ThijmenL98** [comment]
Not sure if you are still active, but I'm having issues with M1's scaling, somehow my application which is about 1900x1000 pixels in size approximately, is blown out of proportion and unable to be scaled down on an M1 with a resolution of 2560x1600. I'm unsure why the M1 scales the application up so much.

**Comment by latop2604** [comment]
Hello, any news about arm64 support and also AOT?

I tried with some friend's mac M1 but it does not seem to work with my app with rid osx-x64 with and without AOT.
The intel version work well with rosetta but not with AOT. (the windows version is ok with AOT)
(I don't have mac myself so I can't build the app on it. I do it in linux with crosscompilation)

Do you know if it's a known issue or if it should be ok?

**Reply by BAndysc** [reply]
I believe you can't build mac version on a non macos. I had the same issue and building the app on mac solved it, the app runs without any problem on other macs. Pretty sure this is macOS thing, not related to Avalonia.

**Reply by kekekeks** [reply]
You need pretty recent builds of .NET for NativeAOT to work with OSX, especially if you want your app to be notarized. Hopefully .NET 8.0 release would include required changes.

**Reply by kekekeks** [reply]
Also, cross-compiling might be a problem since Linux clang doesn't exactly match one shipped with Xcode and doesn't include macOS SDK
